daemon/commands.c

```diff
@@ -0,0 +1,539 @@
+    if (cmd_ctx->message)
+        freez(cmd_ctx->message);
+}
+
+static void schedule_command(uv_work_t *req)
+{
+    struct command_context *cmd_ctx = req->data;
+
+    cmd_ctx->status = execute_command(cmd_ctx->idx, cmd_ctx->args, &cmd_ctx->message);
+}
+
+static void parse_commands(uv_stream_t *client)
+{
+    char *message = NULL, *pos;
+    cmd_t i;
+    cmd_status_t status;
+    struct command_context *cmd_ctx;
+
+    status = CMD_STATUS_FAILURE;
+
+    for (i = 0 ; i < CMD_TOTAL_COMMANDS ; ++i) {
+        pos = strstr(command_string, command_info_array[i].cmd_str);
+        if (pos) {
+            cmd_ctx = mallocz(sizeof(*cmd_ctx));
+            cmd_ctx->work.data = cmd_ctx;
+            cmd_ctx->client = client;
+            cmd_ctx->idx = i;
+            cmd_ctx->args = pos + strlen(command_info_array[i].cmd_str);
+            cmd_ctx->message = NULL;
+
+            assert(0 == uv_queue_work(loop, &cmd_ctx->work, schedule_command, after_schedule_command));
+            break;
+        }
+    }
+    if (CMD_TOTAL_COMMANDS == i) {
+        /* no command found */
+        message = strdupz("Illegal command. Please type \"help\" for instructions.");
+        send_command_reply(client, status, message);
+        freez(message);
+    }
+}
+
+static void pipe_read_cb(uv_stream_t *client, ssize_t nread, const uv_buf_t *buf)
+{
+    if (0 == nread) {
+        info("%s: Zero bytes read by command pipe.", __func__);
+    } else if (UV_EOF == nread) {
+        info("EOF found in command pipe.");
+        parse_commands(client);
+    } else if (nread < 0) {
+        error("%s: %s", __func__, uv_strerror(nread));
+    }
+
+    if (nread < 0) { /* stop stream due to EOF or error */
+        (void)uv_read_stop((uv_stream_t *)client);
+    } else if (nread) {
+        size_t to_copy;
+
+        to_copy = MIN(buf->len, MAX_COMMAND_LENGTH - 1 - command_string_size);
+        strncpyz(command_string + command_string_size, buf->base, to_copy);
+    }
+    if (buf && buf->len) {
+        freez(buf->base);
+    }
+
+    if (nread < 0 && UV_EOF != nread) {
+        uv_close((uv_handle_t *)client, NULL);
+        freez(client);
+        --clients;
+        info("Command Clients = %u\n", clients);
+    }
+}
+
+static void alloc_cb(uv_handle_t *handle, size_t suggested_size, uv_buf_t *buf)
+{
+    (void)handle;
+
+    buf->base = mallocz(suggested_size);
+    buf->len = suggested_size;
+}
+
+static void connection_cb(uv_stream_t *server, int status)
+{
+    int ret;
+    uv_pipe_t *client;
+    assert(status == 0);
+
+    client = mallocz(sizeof(*client));
+    ret = uv_pipe_init(server->loop, client, 1);
+    if (ret) {
+        error("uv_pipe_init(): %s", uv_strerror(ret));
+        freez(client);
+        return;
+    }
+    ret = uv_accept(server, (uv_stream_t *)client);
+    if (ret) {
+        error("uv_accept(): %s", uv_strerror(ret));
+        uv_close((uv_handle_t *)client, NULL);
+        freez(client);
+        return;
+    }
+
+    ++clients;
```

Comment:
I didn't think that we supported multiple client connections - either this should trigger an error or we should separate state for multiple connections.

Reply:
We have separate state indeed.